### PR TITLE
New API implementation

### PR DESCRIPTION
**Pull request by zshipko**

### CODE REVIEW

File: src/irmin_graphql.ml [hunk 1]
```diff
@@ -8,214 +8,394 @@ module type S = sig
...
+            let info = mk_info i in
+            List.fold_right (fun x tree ->
+                match x with
+                | Ok (k, v) -> (tree >>= fun tree -> Store.Tree.add tree k v)
+                | _ -> tree
+              ) items (Store.tree t) >>= fun tree ->
+            Store.set_tree t Store.Key.empty tree ~info >>= fun _ ->
+            Store.Head.find t >>=
+            Lwt.return_ok
+          )
```
Comment (andreas) [inline src/irmin_graphql.ml]: Compared to the earlier design of the schema, this allows setting multiple key/values. What were your thoughts on that?

File: src/irmin_graphql.ml [hunk 2]
```diff
@@ -8,214 +8,394 @@ module type S = sig
...
+        ~resolve:(fun _ _src branch keys i ->
+            mk_branch (Store.repo s) branch >>= fun t ->
+            let info = mk_info i in
+            List.fold_right (fun k tree ->
+                tree >>= fun tree -> Store.Tree.remove tree k
+              ) keys (Store.tree t) >>= fun tree ->
+            Store.set_tree t Store.Key.empty tree ~info >>= fun _ ->
+            Store.Head.find t >>=
+            Lwt.return_ok
+          )
```
Comment (andreas) [inline src/irmin_graphql.ml]: As above: remove one vs remove many?

File: bin/main.ml [hunk 3]
```diff
@@ -1,18 +1,19 @@
 open Lwt.Infix
 
-module Store = Irmin_unix.Git.FS.Make
+module Store = Irmin_unix.Git.Make
+  (Irmin_unix.Git.G)
```
Comment (andreas) [inline bin/main.ml]: It seems this requires the development version of irmin packages to work (similarly for the changes below in this file).

**Comment by andreas** [comment]
For reference, here's what the current schema looks like according to `gqltools`:

```graphql
schema {
  query: query
  mutation: mutation
}

type Branch {
  name: String!
  head: Commit
  get(key: Key!): String
  lca(commit: CommitHash!): [Commit!]!
}

scalar BranchName

type Commit {
  node: Node!
  parents: [Commit!]!
  info: Info!
  hash: String!
}

scalar CommitHash

type Contents {
  key: String!
  value: String
}

type Info {
  date: String!
  author: String!
  message: String!
}

input InfoInput {
  message: String
  author: String
}

input ItemInput {
  value: Value!
  key: String!
}

scalar Key

type mutation {
  clone(remote: Remote!, branch: BranchName): Commit!
  set(info: InfoInput, items: [ItemInput!]!, branch: BranchName): Commit
  remove(info: InfoInput, keys: [Key!]!, branch: BranchName): Commit
  merge(info: InfoInput, from: BranchName!, branch: BranchName): Commit
  push(remote: Remote!, branch: BranchName): String
  pull(remote: Remote!, branch: BranchName): Commit
  revert(commit: CommitHash!, branch: BranchName): Commit
}

type Node {
  key: String!
  get(key: Step!): Node
  tree: [Tree!]!
}

type query {
  master: Branch
  branch(name: BranchName!): Branch
}

scalar Remote

scalar Step

union Tree = Contents | Node

scalar Value
```

File: bin/main.ml (hunk 3 above)
Comment (zshipko) [reply bin/main.ml]: Sorry about that! I will need to update this once we are able to finalize the API.

File: src/irmin_graphql.ml [hunk 4]
```diff
@@ -5,217 +5,440 @@ module type S = sig
...
+    let key = Schema.Arg.(
+        scalar "Key" ~coerce
       )
-    )
 
-    let branch = Schema.Arg.(scalar "BranchName"
-      ~coerce:(function
-        | `String s -> Store.Branch.of_string s |> of_irmin_result
-        | _ -> Error "Branch only accepts strings"
+
```
Comment (andreas) [inline src/irmin_graphql.ml]: Nit: double blankline

File: bin/main.ml [hunk 5]
```diff
@@ -1,20 +1,36 @@
 open Lwt.Infix
+open Cmdliner
```
Comment (andreas) [inline bin/main.ml]: This is a lot nicer with `Cmdliner` 😄

I suppose `cmdliner` should be added as a dependency in `irmin-graphql.opam` and `bin/jbuild`?

**Comment by zshipko** [comment]
Great! I will look into making some of the changes you mentioned in your latest review but will hold off on merging this until the `Irmin_unix.Git` changes have been merged (which should be very soon).

**Comment by zshipko** [comment]
After taking another look at the `get`/`set` functions I am having a hard time figuring out how to make all mutations share a transaction. It would be really nice if all modifications could be done in a single transaction, that's initially why I made these functions take a list of items instead of single key/value pairs, however if every mutation can use the same transaction that would be ideal!

 Did you have an idea of how to implement that? I wasn't able to find a good way to pass context around to each mutation.

**Comment by andreas** [comment]
The `Graphql_lwt.Schema.schema` type is parameterized over a context type `'ctx`. This value is constructed for every request by the `~ctx`-function passed to `Graphql_lwt.Server.start`. All resolve functions accept this value as the first parameter.

At first glance, my idea would be to store the current transaction in this context value, and allow it to be lazily instantiated. I haven't worked out all the details though, in particular when the transaction would be committed, or whether Irmin transactions work across multiple branches (which the current GraphQL mutations allow). Solving the former may require some modifications to `Graphql_lwt.Server` to add appropriate "hook points".

Example with some imaginary `Transaction`-module:

```ocaml
open Graphql_lwt

type ctx = { tx : Transaction.t Lazy.t }

let schema = Schema.(schema [
  field "set"
    ~typ:(non_null string)
    ~args:Arg.[
      arg "key" ~typ:(non_null string);
      arg "value" ~typ:(non_null string)
    ]
    ~resolve:(fun ctx () key value -> Transaction.set Lazy.(force ctx.tx) key value
])

Server.start schema ~ctx:(fun req -> { tx = lazy (Transaction.create ()) } ) schema
```

Another idea would be to add a new mutation field `transaction` like this:

```graphql
enum TRANSACTION_CHANGE_TYPE = {
  SET
  REMOVE
}

input TransactionChangeInput {
  type: TRANSACTION_CHANGE_TYPE
  key: String!
  value: String # this field is nonsensical when type is REMOVE
  info: InfoInput
}

type Mutation {
  transaction(changes: [TransactionChangeInput!]!, branch: Branch): Tree
  # ...
}
```

**Comment by andreas** [comment]
@zshipko: Based on your most recent commit, it seems like you're going for a variation of the `transaction`-mutation field. On second thought, I would propose to stick to single-key `set` and `remove` and ignore transactions in this PR -- just for the sake of not accumulation too many things on this branch. We can then address transactions in a separate PR and give that more careful thought (possibly involving @samoht and other Irmin users).

**Comment by zshipko** [comment]
@andreas Fair enough, I was a little curious what the transaction mutation might look like but was unhappy with what I came up with. I will remove `transaction` (for now, at least) and merge the other changes when I get the chance!

**Pull request merged**